cms/admin/settingsadmin.py

```diff
@@ -97,6 +97,10 @@ def get_toolbar(self, request):
         request = copy.copy(request)
         request.GET = data
         request.current_page = current_page
+        try:
+            request.resolver_match = resolve(origin_url.path)
+        except Resolver404:
```

Comment:
What is the scenario for the 404 being thrown?

There should be a unit test that protects the need for this, it will also expose the scenario where the 404 occurs.

Reply:
Someone could type an url into the browser with a typo in the get parameter `cms_path`.  This will cause the url resolver to fail, say `http://localhost:8000/admin/cms/usersettings/cms-toolbar/?placeholders[]=3&placeholders[]=13&placeholders[]=6&obj_id=7&obj_type=cms.pagecontent&cms_path=%2Fadmin%2Fcms%2Fplaceholder%2Fobject%2F10%2Feat_it%2F7%2F`. The `eat_it` part will not resolve.

Also, code symmetry with https://github.com/django-cms/django-cms/blob/914558d283c197b4035ae7e1a084860f486c9429/cms/toolbar/toolbar.py#L39-L47

Note that line 40/41 prepared for exactly the behavior proposed here.

I think I can add a test for this.